ur_controllers/src/force_torque_sensor_controller.cpp

```diff
@@ -0,0 +1,180 @@
+    {
+      tVec.set__y(state_interface.get_value());
+      continue;
+    }
+
+    if (state_interface.get_interface_name() == "tz")
+    {
+      tVec.set__z(state_interface.get_value());
+      continue;
+    }
+  }
+
+  // TODO set frame_id as parameter --> it includes tf listener within controller
+  wrench_state_msg_.header.stamp = get_node()->get_clock()->now();
+  wrench_state_msg_.header.frame_id = "tool0";
+
+  // update wrench state message
+  wrench_state_msg_.wrench.set__force(fVec);
+  wrench_state_msg_.wrench.set__torque(tVec);
+
+  // publish
+  wrench_state_publisher_->publish(wrench_state_msg_);
+
+  return controller_interface::return_type::SUCCESS;
+}
+
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+ForceTorqueStateController::on_configure(const rclcpp_lifecycle::State& /*previous_state*/)
+{
+  try
+  {
+    // TODO make topic name a parameter
+    wrench_state_publisher_ =
+        get_node()->create_publisher<geometry_msgs::msg::WrenchStamped>("ft_data", rclcpp::SystemDefaultsQoS());
+  }
+  catch (...)
+  {
+    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR;
+  }
+  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+}
+
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+ForceTorqueStateController::on_activate(const rclcpp_lifecycle::State& /*previous_state*/)
+{
+  if (!init_sensor_data())
+  {
+    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR;
+  }
+
+  init_sensor_state_msg();
+
+  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+}
+
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+ForceTorqueStateController::on_deactivate(const rclcpp_lifecycle::State& /*previous_state*/)
+{
+  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+}
+
+template <typename T>
+bool has_any_key(const std::unordered_map<std::string, T>& map, const std::vector<std::string>& keys)
+{
+  bool found_key = false;
+  for (const auto& key_item : map)
+  {
+    const auto& key = key_item.first;
+    if (std::find(keys.cbegin(), keys.cend(), key) != keys.cend())
+    {
+      found_key = true;
+      break;
+    }
+  }
+  return found_key;
+}
+
+bool ForceTorqueStateController::init_sensor_data()
+{
+  bool has_fts_sensor = false;
+  // loop in reverse order, this maintains the order of values at retrieval time
+  for (auto si = state_interfaces_.crbegin(); si != state_interfaces_.crend(); si++)
+  {
+    // TODO remove hardcoded naming
+    if (si->get_name() == "tcp_fts_sensor")
+    {
+      has_fts_sensor = true;
+      axis_val_names_.push_back(si->get_interface_name());
+    }
+  }
+  return has_fts_sensor;
+}
+
+void ForceTorqueStateController::init_sensor_state_msg()
+{
+  // default initialization for wrenchstamped state message
+}
```

Comment:
This function if ported from ROS1, gut in ROS2 is not needed because CM is assigning interface and if none exists, then you get error.
```suggestion
```